esmvaltool/diag_scripts/gwls/plot_gwl_exceedance_mm_stats.py

```diff
@@ -0,0 +1,276 @@
+                "Model: %s, Exp : %s start year: %s, endyear: %s",
+                dataset,
+                el["exp"],
+                str(start_year),
+                str(end_year),
+            )
+            # Start year Jan 1st (1, 1) and end_year Dec 31st (12, 31)
+            cube = extract_time(cube, start_year, 1, 1, end_year, 12, 31)
+            cube = cube.collapsed("time", iris.analysis.MEAN)
+            cubes.append(cube)
+    return cubes
+
+
+def calculate_mm_stats(cubes, mean_file, stdev_file):
+    """Calculate mean and standard deviation from merged mm cubes."""
+    # find index of time coord so this can be unified across data sets
+    # before merging as the time data points are different across models/scenarios
+    index = 0
+    for coord in cubes[0].aux_coords:
+        if coord.standard_name == "time":
+            break
+        index = index + 1
+    for c in cubes[1:]:
+        c.remove_coord("time")
+        c.add_aux_coord(cubes[0].aux_coords[index])
+
+    if len(cubes) == 0:
+        logger.info("No model instances exceed this GWL.")
+    elif len(cubes) == 1:
+        iris.save(cubes[0], mean_file)
+        logger.info("No standard deviation calculated for a single instance.")
+    else:
+        mm_cube = cubes.merge_cube()
+        mm_mean_cube = mm_cube.collapsed(["cube_label"], iris.analysis.MEAN)
+        mm_stdev_cube = mm_cube.collapsed(
+            ["cube_label"], iris.analysis.STD_DEV
+        )
+        iris.save(mm_mean_cube, mean_file)
+        iris.save(mm_stdev_cube, stdev_file)
+
+
+def main(cfg):
+    """Execute GWL statistics and plot."""
+    input_data = cfg["input_data"].values()
+    window_size = cfg["window_size"]
+    gwls = cfg["gwls"]
+    pattern = cfg["pattern"]
+    cmap_mean_str = cfg["quickplot"]["cmap_mean"]
+    cmap_stdev_str = cfg["quickplot"]["cmap_stdev"]
+    mean_level_params = cfg["quickplot"]["mean_level_params"]
+    stdev_level_params = cfg["quickplot"]["stdev_level_params"]
+
+    gwl_filename = io.get_ancestor_file(cfg, pattern)
+    logger.info("GWL exceedance years file is %s", gwl_filename)
+    gwl_df = pd.read_csv(gwl_filename)
+    sep = "_"
+
+    projects = []
+    for data in input_data:
+        # select by by project
+        if data["project"] not in projects:
+            projects.append(data["project"])
+
+    logger.info("List of Projects: %s", projects)
+
+    for project in projects:
+        # get all data sets for project across experiments
+        project_data = select_metadata(input_data, project=project)
+        for gwl in gwls:
+            gwl_subset_df = gwl_df[gwl_df["GWL"] == gwl]
+            if gwl_subset_df.shape[0] > 0:
+                logger.info(
+                    "Calculating means and standard deviations for GWL: %s",
+                    str(gwl),
+                )
+                filename = sep.join([project, "mm_mean", str(gwl)]) + ".nc"
+                mean_file = os.path.join(cfg["work_dir"], filename)
```

Comment:
You could also use the function [`esmvaltool.diag_scripts.shared.get_diagnostic_filename`](https://docs.esmvaltool.org/en/latest/api/esmvaltool.diag_scripts.shared.html#esmvaltool.diag_scripts.shared.get_diagnostic_filename) instead of writing your own, that should simplify the code a bit. For example: `get_diagnostic_filename(f"mm_mean{sep}{gwl}", cfg)` should give you a good path.

Similarly, for the plots you could use [`esmvaltool.diag_scripts.shared.get_plot_filename`](https://docs.esmvaltool.org/en/latest/api/esmvaltool.diag_scripts.shared.html#esmvaltool.diag_scripts.shared.get_plot_filename)